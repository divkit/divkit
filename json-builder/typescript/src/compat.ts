import { DivLinearGradient } from './generated';

/**
 * @deprecated use DivLinearGradient instead.
 */
export class DivGradientBackground extends DivLinearGradient {}
