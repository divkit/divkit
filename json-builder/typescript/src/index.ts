export * from './compat';
export * from './div-utils';
export * from './expression';
export * from './generated';
export * from './helper';
export * from './safe-expression';
export * from './template';
export * from './template-helper';
export * from './template-helper-deps';
export * from './template-helper-memo';
export * from './template-helper-rewrite-names';
export * from './template-helper-rewrite-refs';
export * from './template-helper-version';
