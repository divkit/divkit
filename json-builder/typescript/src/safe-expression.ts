const expr = Symbol();

export interface DivExpression {
    [expr]: string;
}

export class SafeDivExpression implements DivExpression {
    public [expr]: string;

    public constructor(expression: string) {
        this[expr] = expression;
    }

    public toJSON(): string {
        return this[expr];
    }

    public toString(): string {
        return this[expr];
    }
}
