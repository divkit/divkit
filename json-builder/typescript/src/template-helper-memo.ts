import { ITemplates, TemplateBlock } from './template';
import { TemplateHelper } from './template-helper';

export interface ThelperWithMemo<T extends ITemplates> {
    thelper: TemplateHelper<T>;
    used: Set<string>;
}

/**
 * templateHelper with memoization of templates names for later
 * selection of used templates. It memoizes only explicitly used
 * names without tracking dependencies.
 * @param options
 * @param options.customName adds templates names mapping.
 * Instead of original template name adjust template's type to custom value
 * returned by customName. Can be used for templates versioning.
 */
export function thelperWithMemo<T extends ITemplates>(options?: {
    customName?: (x: string) => string;
}): ThelperWithMemo<T> {
    const used = new Set<string>();

    const proxyHandler: ProxyHandler<{ [k: string]: (o: unknown) => TemplateBlock<string> }> = {
        get(_, templateName: string) {
            const templateRenamed = options?.customName ? options.customName(templateName) : templateName;
            used.add(templateName);

            return (props: any) => ({ ...props, type: templateRenamed });
        },
    };

    const thelper = new Proxy({}, proxyHandler) as TemplateHelper<T>;

    return { used, thelper };
}
