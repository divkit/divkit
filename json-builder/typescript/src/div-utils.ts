import { DivFixedSize, DivMatchParentSize, DivSizeUnit, DivWrapContentSize } from './generated';
import { DivExpression } from './safe-expression';

export function fixed(value: number | DivExpression, unit: DivSizeUnit | undefined = undefined): DivFixedSize {
    if (unit === undefined) {
        return new DivFixedSize({ value: value });
    }
    return new DivFixedSize({ unit: unit, value: value });
}

export function matchParent(): DivMatchParentSize {
    return new DivMatchParentSize();
}

export function weighted(weight: number): DivMatchParentSize {
    return new DivMatchParentSize({ weight: weight });
}

export function wrapContent(): DivWrapContentSize {
    return new DivWrapContentSize();
}
