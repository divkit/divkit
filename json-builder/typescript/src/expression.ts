import { DivExpression, SafeDivExpression } from './safe-expression';

export function expression(expression: string): DivExpression {
    return new SafeDivExpression(expression);
}

const replacer: Record<string, string> = {
    '\\': '\\\\',
    '@{': '\\@{',
};

export function escapeExpression(str: string): string {
    return str.replace(/\\|(@{)/g, (full) => {
        return replacer[full];
    });
}
