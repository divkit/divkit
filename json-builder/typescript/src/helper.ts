import { ITemplates } from './template';
import { SafeDivExpression } from './safe-expression';

export type NonEmptyArray<T> = T[];

export type Exact<TBase, TExt extends TBase> = TExt extends unknown ? Exactly<TBase, TExt> : never;

type Exactly<TBase, TExt extends TBase> = {
    [K in keyof TExt]: K extends keyof TBase ? TExt[K] : never;
};

export type IntBoolean = 1 | 0 | true | false;

/**
 * DFS for a js object with callback on each leaf
 * @param tree js object
 * @param nodeAction callback for each leaf
 */
export function treeWalkDFS(tree: unknown, nodeAction: (x: unknown, path: string[]) => void): void {
    const stack: [unknown, string[]][] = [[tree, []]];

    while (stack.length) {
        const [node, path] = stack.pop() as [unknown, string[]];
        nodeAction(node, path);

        if (typeof node === 'object' && node !== null) {
            const list = Object.entries(node as { [k: string]: unknown }).map(
                ([key, obj]) => [obj, [...path, key]] as [unknown, string[]],
            );

            for (let i = list.length - 1; i >= 0; --i) {
                stack.push(list[i]);
            }
        }
    }
}

export function copyTemplates<T extends ITemplates>(templates: T): T {
    const copy: ITemplates = {};

    const placeHolder = (node: unknown): unknown => {
        if (!node || ['string', 'number', 'boolean'].includes(typeof node) || node instanceof SafeDivExpression) {
            return node;
        }

        if (Array.isArray(node)) {
            return [];
        }

        return Object.create(Object.getPrototypeOf(node));
    };

    const clone = (node: unknown, path: string[]): void => {
        let parentPointer: object = copy;

        path = [...path];
        while (path.length > 1) {
            parentPointer = parentPointer[path.shift() as string];
        }

        const cur = path[0];

        if (!cur) {
            return;
        }

        parentPointer[cur] = placeHolder(node);
    };

    treeWalkDFS(templates, clone);

    return copy as T;
}
