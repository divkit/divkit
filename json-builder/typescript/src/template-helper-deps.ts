import { Div } from './generated/Div';
import { treeWalkDFS } from './helper';
import { ITemplates, TemplateBlock } from './template';

/**
 * Search for a used templates list
 * @param template
 */
function findPlainDeps(template: Div): Set<string> {
    const deps = new Set<string>();

    treeWalkDFS(template, (node) => {
        if (node instanceof TemplateBlock) {
            deps.add(node.type);
        }
    });

    return deps;
}

const inProgress = 1;
const isFinished = 2;

export type TemplateResolvedAction<T> = (props: {
    name: string;
    template: Div;
    plainDeps: string[];
    depsResolved: { [k: string]: T };
}) => T;

/**
 * Templates walker. It resolves template dependencies on each step, and also can run a callback after that
 * @param templates
 * @param resolvedAction Callback for the template. It will be called after resolving dependencies
 * @param depsResolved Deps resolve cache, for example, for common templates
 */
export function runResolveDeps<T>(
    templates: ITemplates,
    resolvedAction: TemplateResolvedAction<T>,
    depsResolved: { [k: string]: T } = {},
): { [k: string]: T } {
    depsResolved = { ...depsResolved };
    const progress: { [k: string]: undefined | typeof inProgress | typeof isFinished } = {};

    for (const k of Object.keys(depsResolved)) {
        progress[k] = isFinished;
    }

    const stack = Object.keys(templates);

    while (stack.length) {
        const name = stack.pop() as string;

        if (progress[name] === isFinished) {
            continue;
        }

        const plainDeps = [...findPlainDeps(templates[name])];
        const unresolvedDeps = plainDeps.filter((dep) => !(progress[dep] === isFinished));
        const unsolvableDeps = unresolvedDeps.filter((depName) => !templates[depName]);

        if (unsolvableDeps.length !== 0) {
            throw new Error(
                `template '${name}' unsolvable dependencies: ${unsolvableDeps.map((b) => `'${b}'`).join(',')}`,
            );
        }

        if (unresolvedDeps.length === 0) {
            const template = templates[name];
            depsResolved[name] = resolvedAction({ name, template, plainDeps, depsResolved });
            progress[name] = isFinished;
        } else if (progress[name] === inProgress) {
            throw new Error(`template ${name}: cyclic depdendencies`);
        } else {
            progress[name] = inProgress;
            stack.push(name, ...unresolvedDeps);
        }
    }

    return depsResolved as { [k: string]: T };
}

/**
 * Search for a templates dependencies
 * @param templates
 * @param depsResolved Deps resolution cache
 * @returns Map with deps for each template
 */
export function templatesDepsMap(
    templates: ITemplates,
    depsResolved: { [k: string]: string[] } = {},
): { [k: string]: string[] } {
    return runResolveDeps(
        templates,
        ({ name, plainDeps, depsResolved }) => {
            const deps = new Set([name]);

            for (const plainDep of plainDeps) {
                for (const dep of depsResolved[plainDep]) {
                    deps.add(dep);
                }
            }

            return [...deps];
        },
        depsResolved,
    );
}
