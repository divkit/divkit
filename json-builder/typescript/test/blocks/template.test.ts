import { divCard, DivContainer, DivImage, DivText, fixed, IDivData, reference, rewriteRefs, template } from '../../src';

describe('Div templates test', (): void => {
    it('should create simple template', (): void => {
        const templates = {
            title_text: new DivText({
                text: reference('title_text'),
            }),
        };

        expect(rewriteRefs(templates)).toMatchSnapshot();
    });

    it('should create template with title menu and footer', (): void => {
        const templates = {
            title_text: new DivText({
                text: reference('text'),
                paddings: {
                    left: 16,
                    top: 12,
                    right: 16,
                },
                font_size: 18,
                font_weight: 'medium',
                line_height: 24,
                text_color: '#CC000000',
            }),
            title_menu: new DivImage({
                width: fixed(44),
                height: fixed(44),
                image_url: 'https://i.imgur.com/qNJQKU8.png',
            }),
            footer: new DivText({
                text: reference('footer_text'),
                font_size: 12,
                line_height: 16,
                text_color: '#80000000',
                margins: {
                    left: 16,
                    right: 16,
                },
                paddings: {
                    bottom: 12,
                    top: 12,
                },
                action: reference('footer_action_link'),
            }),
        };

        expect({ templates: rewriteRefs(templates) }).toMatchSnapshot();
    });

    it('should create card with template which has more than one base template block', (): void => {
        const templates = {
            card: template('A', {
                alpha: reference('opacity'),
            }),
            A: template('B'),
            B: template('C', {
                orientation: reference('orient'),
            }),
            C: new DivContainer({
                items: reference('content'),
            }),
        };

        const divData: IDivData = {
            log_id: 'id',
            states: [
                {
                    state_id: 1,
                    div: template('card', {
                        opacity: 0.5,
                        orient: 'horizontal',
                        content: [new DivText({ text: 'text' })],
                        margins: {
                            top: 1,
                        },
                    }),
                },
            ],
        };

        expect(divCard(rewriteRefs(templates), divData)).toMatchSnapshot();
    });

    it('should create nested template and card with new property', (): void => {
        const templates = {
            district2: new DivContainer({
                items: [
                    template('district2_foot_text', {
                        text: reference('comments_cnt'),
                        my_end: reference('my_first_end'),
                    }),
                    template('district2_foot_text', {
                        text: reference('likes_cnt'),
                        my_end: 5,
                    }),
                ],
                column_span: 1,
            }),
            district2_foot_text: new DivText({
                text: reference('text'),
                font_size: 13,
                line_height: 16,
                text_color: '#999',
                width: {
                    type: 'wrap_content',
                },
                paddings: {
                    left: 5,
                },
                alignment_horizontal: 'right',
                ranges: [
                    {
                        start: 0,
                        end: reference('my_end'),
                        text_color: '#333',
                    },
                ],
            }),
        };

        const divData: IDivData = {
            log_id: 'id',
            states: [
                {
                    state_id: 1,
                    div: template('district2', {
                        comments_cnt: 'some.comments.cnt',
                        likes_cnt: 'some.likes.cnt',
                        my_first_end: 5,
                    }),
                },
            ],
        };

        expect(divCard(rewriteRefs(templates), divData)).toMatchSnapshot();
    });
});
