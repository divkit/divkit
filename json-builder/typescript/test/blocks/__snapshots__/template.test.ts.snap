// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Div templates test should create card with template which has more than one base template block 1`] = `
{
  "templates": {
    "card": {
      "type": "A",
      "$alpha": "opacity"
    },
    "A": {
      "type": "B"
    },
    "B": {
      "type": "C",
      "$orientation": "orient"
    },
    "C": {
      "type": "container",
      "$items": "content"
    }
  },
  "card": {
    "log_id": "id",
    "states": [
      {
        "state_id": 1,
        "div": {
          "type": "card",
          "opacity": 0.5,
          "orient": "horizontal",
          "content": [
            {
              "type": "text",
              "text": "text"
            }
          ],
          "margins": {
            "top": 1
          }
        }
      }
    ]
  }
}
`;

exports[`Div templates test should create nested template and card with new property 1`] = `
{
  "templates": {
    "district2": {
      "type": "container",
      "column_span": 1,
      "items": [
        {
          "type": "district2_foot_text",
          "$text": "comments_cnt",
          "$my_end": "my_first_end"
        },
        {
          "type": "district2_foot_text",
          "my_end": 5,
          "$text": "likes_cnt"
        }
      ]
    },
    "district2_foot_text": {
      "type": "text",
      "alignment_horizontal": "right",
      "font_size": 13,
      "line_height": 16,
      "paddings": {
        "left": 5
      },
      "ranges": [
        {
          "start": 0,
          "text_color": "#333",
          "$end": "my_end"
        }
      ],
      "text_color": "#999",
      "width": {
        "type": "wrap_content"
      }
    }
  },
  "card": {
    "log_id": "id",
    "states": [
      {
        "state_id": 1,
        "div": {
          "type": "district2",
          "comments_cnt": "some.comments.cnt",
          "likes_cnt": "some.likes.cnt",
          "my_first_end": 5
        }
      }
    ]
  }
}
`;

exports[`Div templates test should create simple template 1`] = `
{
  "title_text": {
    "type": "text",
    "$text": "title_text"
  }
}
`;

exports[`Div templates test should create template with title menu and footer 1`] = `
{
  "templates": {
    "title_text": {
      "type": "text",
      "font_size": 18,
      "font_weight": "medium",
      "line_height": 24,
      "paddings": {
        "left": 16,
        "top": 12,
        "right": 16
      },
      "text_color": "#CC000000"
    },
    "title_menu": {
      "type": "image",
      "height": {
        "type": "fixed",
        "value": 44
      },
      "image_url": "https://i.imgur.com/qNJQKU8.png",
      "width": {
        "type": "fixed",
        "value": 44
      }
    },
    "footer": {
      "type": "text",
      "font_size": 12,
      "line_height": 16,
      "margins": {
        "left": 16,
        "right": 16
      },
      "paddings": {
        "bottom": 12,
        "top": 12
      },
      "text_color": "#80000000",
      "$action": "footer_action_link",
      "$text": "footer_text"
    }
  }
}
`;
